File: scripts/update_scholar_metrics.py
#!/usr/bin/env python3
import json
import sys
from scholarly import scholarly, ProxyGenerator

SCHOLAR_ID = "ox2t_YIAAAAJ"

# Try to initialize free proxies, but don’t fail if that API changed
pg = ProxyGenerator()
try:
    pg.FreeProxies()               
    scholarly.use_proxy(pg)
    print("[info] Using FreeProxies()")
except Exception as e:
    print(f"[warning] could not init FreeProxies(): {e}", file=sys.stderr)
    # continue without proxy

def fetch_metrics(user_id):
    author = scholarly.search_author_id(user_id)
    author = scholarly.fill(author, sections=[])
    return {
        "citations": author.get("citedby", 0),
        "h_index":   author.get("hindex", 0),
        "i10_index": author.get("i10index", 0),
    }

if __name__ == "__main__":
    try:
        metrics = fetch_metrics(SCHOLAR_ID)
    except Exception as e:
        print(f"[warning] could not fetch metrics: {e}", file=sys.stderr)
        sys.exit(0)

    with open("scholar-metrics.json", "w") as f:
        json.dump(metrics, f, indent=2)
    print("Updated scholar‑metrics.json:", metrics)

File: scripts/update_scholar_metrics.R
#!/usr/bin/env Rscript

# URL of your Google Scholar profile
SCHOLAR_URL <- "https://scholar.google.com/citations?user=ox2t_YIAAAAJ&hl=en"
OUT_FILE    <- "scholar-metrics.json"

# Fetch the HTML
html <- tryCatch({
  paste(readLines(SCHOLAR_URL, warn = FALSE), collapse = "\n")
}, error = function(e) {
  warning("Failed to download page: ", e$message)
  quit(status = 0)  # don’t fail the workflow
})

# Extract the first three <td>…</td> with numbers (citations, h-index, i10-index)
raw_cells <- regmatches(
  html,
  gregexpr("<td[^>]*>[0-9,]+</td>", html, perl = TRUE)
)[[1]]

if (length(raw_cells) < 3) {
  warning("Unexpected page structure; found ", length(raw_cells), " numeric cells")
  quit(status = 0)
}

# Strip tags, commas, convert to integer
nums <- as.integer(gsub(",", "", gsub("<[^>]+>", "", raw_cells[1:3])))

# Build JSON text manually
json_txt <- sprintf(
  "{\n  \"citations\": %d,\n  \"h_index\": %d,\n  \"i10_index\": %d\n}\n",
  nums[1], nums[2], nums[3]
)

# Write out
cat(json_txt, file = OUT_FILE)
message("✔ Updated ", OUT_FILE, ": ", nums[1], "/", nums[2], "/", nums[3])
